src/compute/src/logging/persist.rs

```diff
@@ -0,0 +1,165 @@
+
+    let location = target.persist_location.clone();
+
+    let handles: Option<(WriteHandle<_, _, _, _>, _)> = if active_worker_index == scope.index() {
+        let shard_id = target.data_shard;
+        let persist_client = futures_executor::block_on(async {
+            compute_state
+                .persist_clients
+                .lock()
+                .await
+                .open(location)
+                .await
+        });
+
+        let persist_client = persist_client.expect("Successful connection");
+
+        let (write, read) = futures_executor::block_on(
+            persist_client.open::<SourceData, (), Timestamp, Diff>(shard_id),
+        )
+        .expect("could not open persist shard");
+
+        Some((write, read))
+    } else {
+        None
+    };
+
+    let mut sink = OperatorBuilder::new("Logging Persist Sink".into(), scope.clone());
+    let mut input = sink.new_input(
+        &collection.inner,
+        Exchange::new(move |_| active_worker_index as u64),
+    );
+
+    let shared_frontier = Rc::new(RefCell::new(Antichain::from_elem(
+        TimelyTimestamp::minimum(),
+    )));
+
+    compute_state
+        .reported_frontiers
+        .insert(target_id, Antichain::from_elem(0));
+
+    compute_state
+        .sink_write_frontiers
+        .insert(target_id, Rc::clone(&shared_frontier));
+
+    sink.build_async(
+        scope,
+        move |_capabilities, frontiers, scheduler| async move {
+            let (mut write, read) = match handles {
+                Some(w) => w,
+                None => return,
+            };
+
+            // Delete existing data in shard, by reading at recent_upper - 1
+            // and writing -1s at recent_upper.
+            let recent_upper = write.fetch_recent_upper().await;
+            // If this is true, we can obtain a snapshot. Otherwise, we assume the shard is empty
+            if recent_upper[0] > 0 {
+                let mut initial_flush: Vec<((SourceData, _), _, i64)> = vec![];
+                // Flush the shard
+                let since = Antichain::from_elem(recent_upper[0] - 1);
+                match read.snapshot(since).await {
+                    Ok(mut x) => {
+                        while let Some(result_vec) = x.next().await {
+                            for result in result_vec {
+                                let sd = result.0 .0.unwrap();
+                                let mult = result.2;
+                                initial_flush.push(((sd, ()), recent_upper[0], -mult));
+                            }
+                        }
+                    }
+                    Err(_err) => {
+                        tracing::warn!("Could not flush logging shard!");
+                    }
+                };
+
+                // Use smallest possible new_upper
+                let new_upper = Antichain::from_elem(recent_upper[0] + 1);
+                write
+                    .append(initial_flush.into_iter(), recent_upper, new_upper)
```

Comment:
`append` is only correct for definite data. I _think_ that's true here given that it's a definite computation of data you're reading from persist (which is definite), so I think the only thing you have to worry about is race conditions. i see above only one operator is running this, so that checks out. I assume different compute replicas are then pointed at different shards? if so, this all checks out